File: Day3/solution.js
const fs = require('fs');
const { posix } = require('path');

const forrest = fs
    .readFileSync('./input.txt', 'utf-8')
    .split('\r\n')
    .map((e) => e.split(''));

function checkSlope(xIncrement, yIncrement) {
    let pos = {
        x: 0,
        y: 0,
    };
    let treesEncountered = 0;
    forrestWidth = forrest[0].length;

    while (pos.y < forrest.length) {
        if (forrest[pos.y][pos.x % forrestWidth] === '#') {
            treesEncountered++;
        }
        pos.x += xIncrement;
        pos.y += yIncrement;
    }

    return treesEncountered;
}

// Part 1
console.log('Part 1: ', checkSlope(3, 1));

// Part 2
const slopesToCheck = [
    [1, 1],
    [3, 1],
    [5, 1],
    [7, 1],
    [1, 2],
];
const result = slopesToCheck.reduce((a, c) => {
    let [x, y] = c;
    return a * checkSlope(x,y);
}, 1);
console.log('Part 2: ', result);

File: Day3/solution.py
from functools import reduce

data = open('./input.txt').read().splitlines()


def check_slope(xInc, yInc):
    x = 0
    y = 0
    trees_seen = 0

    while y < len(data):
        if data[y][x % len(data[y])] == '#':
            trees_seen += 1

        x += xInc
        y += yInc

    return trees_seen


print(f'Part 1: {check_slope(3,1)}')
slopes_to_check = [(1, 1), (3, 1), (5, 1), (7, 1), (1, 2)]
print(
    f'Part 2: {reduce((lambda x, y: x*y), [check_slope(a, b) for a, b in slopes_to_check])}')
